packages/context/src/__tests__/unit/context-config.unit.ts

```diff
@@ -0,0 +1,137 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {
+  BindingAddress,
+  BindingKey,
+  ConfigurationResolver,
+  Context,
+  DefaultConfigurationResolver,
+  ResolutionOptions,
+  ValueOrPromise,
+} from '../..';
+
+describe('Context binding configuration', () => {
+  /**
+   * Create a subclass of context so that we can access parents and registry
+   * for assertions
+   */
+  class TestContext extends Context {
+    public configResolver: ConfigurationResolver;
+  }
+
+  let ctx: TestContext;
+  beforeEach(createContext);
+
+  describe('configure()', () => {
+    it('configures options for a binding before it is bound', () => {
+      const bindingForConfig = ctx.configure('foo').to({x: 1});
+      expect(bindingForConfig.key).to.equal(
+        BindingKey.buildKeyForConfig('foo'),
+      );
+      expect(bindingForConfig.tagMap).to.eql({config: 'foo'});
+    });
+
+    it('configures options for a binding after it is bound', () => {
+      ctx.bind('foo').to('bar');
+      const bindingForConfig = ctx.configure('foo').to({x: 1});
+      expect(bindingForConfig.key).to.equal(
+        BindingKey.buildKeyForConfig('foo'),
+      );
+      expect(bindingForConfig.tagMap).to.eql({config: 'foo'});
+    });
+  });
+
+  describe('getConfig()', () => {
+    it('gets config for a binding', async () => {
+      ctx.configure('foo').toDynamicValue(() => Promise.resolve({x: 1}));
+      expect(await ctx.getConfig('foo')).to.eql({x: 1});
+    });
+
+    it('gets local config for a binding', async () => {
+      ctx
+        .configure('foo')
+        .toDynamicValue(() => Promise.resolve({a: {x: 0, y: 0}}));
+      ctx.configure('foo.a').toDynamicValue(() => Promise.resolve({x: 1}));
+      expect(await ctx.getConfig<number>('foo.a', 'x')).to.eql(1);
+      expect(await ctx.getConfig<number>('foo.a', 'y')).to.be.undefined();
+    });
+
+    it('defaults optional to true for config resolution', async () => {
+      // `servers.rest` does not exist yet
+      let server1port = await ctx.getConfig<number>('servers.rest', 'port');
+      expect(server1port).to.be.undefined();
+
+      // Now add `servers.rest`
+      ctx.configure('servers.rest').to({port: 3000});
+      server1port = await ctx.getConfig<number>('servers.rest', 'port');
+      expect(server1port).to.eql(3000);
+    });
+
+    it('throws error if a required config cannot be resolved', async () => {
+      expect(
+        ctx.getConfig('servers.rest', 'host', {
+          optional: false,
+        }),
+      )
+        .to.be.rejectedWith(
+          `Configuration 'servers.rest#host' cannot be resolved`,
```

Comment:
I find this slightly confusing. It's `servers.rest` that could not be resolved, the property accessor `#host` does not play any role here, correct? I would expect to see the message above if the `host` property was not set.